### PR TITLE
feat: event payload implementation

### PR DESCRIPTION
**Pull request by SKannaniOS**
## Description
This PR includes the payload implementation for Track, Screen, and Group events.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Code refactor/optimization

## Implementation Details
Each event contains several common fields in addition to fields unique to that specific event type. This PR implements all the common fields for events. Additionally, it introduces the `RudderOptions` model to manage `integrations` and `customContext` values. The `RudderIdentifyOptions` model is also introduced and will be utilized in future event implementations.

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added the necessary documentation (if appropriate).
- [x] I have ensured that my code follows the project's code style.
- [x] I have checked for potential performance impacts and optimized if necessary.
- [x] I have checked the code for security issues.
- [ ] I have updated the changelog (if required).

## How to test?

To run the sample app, configure the necessary values such as the write key and data plane URL in the AnalyticsApp class. The data plane will receive the Track, Screen, and Group events.

Sample Track Event JSON:
```
{
    "anonymousId": "CC23B8E5-5D97-4225-9E5C-03DF5947F303",
    "channel": "mobile",
    "context": {
        "SK1": {
            "Key1": "Value1"
        },
        "SK2": [
            "value1",
            "value2"
        ],
        "SK3": "Value3",
        "SK4": 1234,
        "SK5": 5678.9,
        "SK6": true
    },
    "event": "Track at 2024-11-14 18:01:51 +0000",
    "integrations": {
        "All": true,
        "SDK": true,
        "Segment": false
    },
    "messageId": "40F9AA0A-33A6-4AAF-A08E-26692F532F56",
    "originalTimeStamp": "2024-11-14T18:01:51.700Z",
    "properties": {
        "key": "value"
    },
    "receivedAt": "2024-11-14T18:02:00.305Z",
    "request_ip": "106.51.148.64",
    "rudderId": "2836452c-f55b-40aa-aef2-b79476455727",
    "sentAt": "2024-11-14T18:01:58.144Z",
    "type": "track"
}
```

Sample Screen Event JSON:
```
{
    "anonymousId": "E30C3220-8ECC-4746-83B7-657F214975D7",
    "channel": "mobile",
    "context": {
        "SK": {
            "Key1": "Value1"
        }
    },
    "event": "Analytics Screen",
    "integrations": {
        "All": true,
        "Segment": false
    },
    "messageId": "458C7801-D471-4039-9CEF-E44D2013FE97",
    "originalTimeStamp": "2024-11-14T18:04:58.558Z",
    "properties": {
        "key": "value",
        "name": "Analytics Screen"
    },
    "receivedAt": "2024-11-14T18:05:09.035Z",
    "request_ip": "106.51.148.64",
    "rudderId": "d11cbd78-d252-4bf6-a29f-6eee75b74220",
    "sentAt": "2024-11-14T18:05:08.134Z",
    "type": "screen"
}
```

Sample Group Event JSON:
```
{
    "anonymousId": "4EC697F2-9BD2-4016-AD2C-7756E4C59CA3",
    "channel": "mobile",
    "context": {
        "SK": {
            "Key1": "Value1"
        }
    },
    "groupId": "group_id",
    "integrations": {
        "All": true,
        "MySDK": false
    },
    "messageId": "F1EB4361-CCDA-40A1-89D1-D098C9EE4F7E",
    "originalTimeStamp": "2024-11-14T18:05:36.376Z",
    "receivedAt": "2024-11-14T18:05:38.283Z",
    "request_ip": "106.51.148.64",
    "rudderId": "75bdfd20-2cd2-4d52-8b7b-88c0cf71660b",
    "sentAt": "2024-11-14T18:05:38.132Z",
    "traits": {
        "anonymousId": "4EC697F2-9BD2-4016-AD2C-7756E4C59CA3",
        "key": "value"
    },
    "type": "group"
}
```

### CODE REVIEW

File: AnalyticsApp/AnalyticsApp/ContentView.swift [hunk 1]
```diff
@@ -6,30 +6,52 @@
...
 struct ContentView: View {
     
     var body: some View {
         VStack {
             HStack {
                 CustomButton(title: "Track") {
-                    AnalyticsManager.shared.analytics?.track(name: "Track at \(Date())", properties: ["key": "value"], options: ["option": "value"])
+                    let option = RudderOptions()
+                    .addIntegration("SDK", isEnabled: true)
+                    .addIntegration("Segment", isEnabled: false)
```
Comment (1abhishekpandey) [inline AnalyticsApp/AnalyticsApp/ContentView.swift]: Please replace it with real values like `Amplitude` or `CleverTap`.
```suggestion
                    .addIntegration("Amplitude", isEnabled: true)
                    .addIntegration("CleverTap", isEnabled: false)
```

File: AnalyticsApp/AnalyticsApp/ContentView.swift [hunk 2]
```diff
@@ -6,30 +6,52 @@
...
+                    .addCustomContext(5678.9, key: "SK5")
+                    .addCustomContext(true, key: "SK6")
+                    
+                    AnalyticsManager.shared.analytics?.track(name: "Track at \(Date())", properties: ["key": "value"], options: option)
                 }
                 
                 CustomButton(title: "Multiple Track") {
+                    let option = RudderOptions()
+                    .addIntegration("SDK", isEnabled: true)
+                    .addIntegration("Segment", isEnabled: false)
```
Comment (1abhishekpandey) [inline AnalyticsApp/AnalyticsApp/ContentView.swift]: Update these keys as well.

File: AnalyticsApp/AnalyticsApp/ContentView.swift [hunk 3]
```diff
@@ -6,30 +6,52 @@
...
+                        .addCustomContext(["Key1": "Value1"], key: "SK")
+                        .addIntegration("Segment", isEnabled: false)
+                    AnalyticsManager.shared.analytics?.screen(name: "Analytics Screen", properties: ["key": "value"], options: option)
                 }
                 
                 CustomButton(title: "Group") {
-                    AnalyticsManager.shared.analytics?.group(id: "group_id", traits: ["key": "value"], options: ["option": "value"])
+                    let option = RudderOptions()
+                        .addCustomContext(["Key1": "Value1"], key: "SK")
+                        .addIntegration("MySDK", isEnabled: false)
```
Comment (1abhishekpandey) [inline AnalyticsApp/AnalyticsApp/ContentView.swift]: Update this key as well.

File: Analytics/Analytics/Events/Message.swift
Comment (1abhishekpandey) [inline Analytics/Analytics/Events/Message.swift:1]: Let's add documentation for all the events i.e., track, screen and group.

Also, consider moving them into a separate class, under some event directory, if seems correct.

File: Analytics/Analytics/Events/MessageOptions.swift [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        return self
+    }
+    
+    @discardableResult
+    public func addExternalId(_ id: ExternalId) -> Self {
+        if self.integrations == nil { self.externalIds = [] }
+        
+        self.externalIds?.append(id)
+        return self
+    }
```
Comment (1abhishekpandey) [inline Analytics/Analytics/Events/MessageOptions.swift]: Let's remove this method. As this is part of the identify event (not the current set of events). Let's not add anything, which is not part of the current scope.

Comment (1abhishekpandey) [reply Analytics/Analytics/Events/MessageOptions.swift]: Remove all the instances of externalIds.

File: Analytics/Analytics/Events/MessageOptions.swift [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+
+// MARK: - RudderIdentifyOptionType
+protocol RudderIdentifyOptionType: RudderOptionType {
+    var externalIds: [ExternalId]? { get }
+}
+
+// MARK: - RudderIdentifyOptions
+public class RudderIdentifyOptions: RudderIdentifyOptionType {
```
Comment (1abhishekpandey) [inline Analytics/Analytics/Events/MessageOptions.swift]: Please remove identify event-related code.

File: Analytics/AnalyticsTests/MessageModuleTests.swift [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+//  MessageModuleTests.swift
+//  AnalyticsTests
+//
+//  Created by Satheesh Kannan on 13/11/24.
+//
+
+import XCTest
+@testable import Analytics
+
+final class MessageModuleTests: XCTestCase {
```
Comment (1abhishekpandey) [inline Analytics/AnalyticsTests/MessageModuleTests.swift:11]: Let's refactor the test classes to follow the Android way to construct and match the exact JSON payload, refer [here](https://github.com/rudderlabs/kotlin-sdk-poc/blob/16e87407d4ef81cc077a385997ad340f2c058787/core/src/test/kotlin/com/rudderstack/kotlin/sdk/internals/models/TrackEventTest.kt#L40) for test class, and [here](https://github.com/rudderlabs/kotlin-sdk-poc/tree/16e87407d4ef81cc077a385997ad340f2c058787/core/src/test/resources/message/track) for the payloads.

File: Analytics/AnalyticsTests/MessageModuleTests.swift [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+                    XCTAssertFalse(track.integrations?.isEmpty ?? true)
+                    XCTAssertFalse(track.sentAt?.isEmpty ?? true)
+                    XCTAssertFalse(track.context?.isEmpty ?? true)
+                }
+            }
+        }
+    }
+    
+    // MARK: - Screen
+    func test_screen_event() {
```
Comment (1abhishekpandey) [inline Analytics/AnalyticsTests/MessageModuleTests.swift:153]: Let's have a separate test class for the screen event.

File: Analytics/AnalyticsTests/MessageModuleTests.swift [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+                    XCTAssertNotNil(screen.integrations)
+                    XCTAssertFalse(screen.integrations?.isEmpty ?? true)
+                    XCTAssertFalse(screen.sentAt?.isEmpty ?? true)
+                    XCTAssertFalse(screen.context?.isEmpty ?? true)
+                }
+            }
+        }
+    }
+    
+    // MARK: - Group
```
Comment (1abhishekpandey) [inline Analytics/AnalyticsTests/MessageModuleTests.swift:291]: Let's have a separate test class for the group event.

File: Analytics/AnalyticsTests/MessageModuleTests.swift (hunk 6 above)
Comment (1abhishekpandey) [inline Analytics/AnalyticsTests/MessageModuleTests.swift:11]: Let's rename this class to be of type track.

File: Analytics/AnalyticsTests/MessageModuleTests.swift [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+                    XCTAssertNotNil(track.integrations)
+                    XCTAssertFalse(track.integrations?.isEmpty ?? true)
+                    XCTAssertFalse(track.sentAt?.isEmpty ?? true)
+                    XCTAssertFalse(track.context?.isEmpty ?? true)
+                }
+            }
+        }
+    }
+    
+    // MARK: - Screen
```
Comment (1abhishekpandey) [inline Analytics/AnalyticsTests/MessageModuleTests.swift:152]: We also need to test the category. Please write a unit test for that.

**Review by 1abhishekpandey — CHANGES_REQUESTED** [review]

File: Analytics/Analytics/Events/Message.swift
Comment (SKannaniOS) [reply Analytics/Analytics/Events/Message.swift:1]: I have added basic documentation for the event models. There’s no need to create separate classes for each event at this stage. If the model size exceeds a certain limit, separate classes will be introduced in a future PR.

File: Analytics/AnalyticsTests/MessageModuleTests.swift (hunk 6 above)
Comment (SKannaniOS) [reply Analytics/AnalyticsTests/MessageModuleTests.swift:11]: Event test cases will be refactored as a next PR.
FYR: https://linear.app/rudderstack/issue/SDK-2662/[ios]-refactor-tests-in-event-prs

File: Analytics/AnalyticsTests/MessageModuleTests.swift (hunk 9 above)
Comment (SKannaniOS) [reply Analytics/AnalyticsTests/MessageModuleTests.swift:152]: Analytics tests will cover this scenario. But anyway, event test cases refactor PR will cover this again.

FYR: [https://linear.app/rudderstack/issue/SDK-2662/[ios]-refactor-tests-in-event-prs](https://linear.app/rudderstack/issue/SDK-2662/%5Bios%5D-refactor-tests-in-event-prs)

File: Analytics/AnalyticsTests/MessageModuleTests.swift (hunk 6 above)
Comment (SKannaniOS) [reply Analytics/AnalyticsTests/MessageModuleTests.swift:11]: Event test cases will be refactored to have each event in its own class file. However, this won’t be implemented in the near future.

File: Analytics/AnalyticsTests/MessageModuleTests.swift (hunk 8 above)
Comment (SKannaniOS) [reply Analytics/AnalyticsTests/MessageModuleTests.swift:291]: Event test cases will be refactored in the next PR, with each event placed in a separate class file.

File: Analytics/AnalyticsTests/MessageModuleTests.swift (hunk 7 above)
Comment (SKannaniOS) [reply Analytics/AnalyticsTests/MessageModuleTests.swift:153]: Event test cases will be refactored in the next PR, with each event placed in a separate class file.

**Review by 1abhishekpandey — APPROVED** [review]

**Review by 1abhishekpandey — APPROVED** [review]

**Review by ChryssaAliferi — APPROVED** [review]
🚀 Amazing! Let's gooo!

**Pull request merged**